File: Depth-Perception-Using-Stereo-Camera/python/obstacle_avoidance.py
"""Real-rig obstacle visualization built on the tested stereo-depth core."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2

from stereo_depth import (
    DEFAULT_CONFIG,
    DEFAULT_MAPS,
    compute_disparity,
    create_matcher,
    disparity_to_depth,
    disparity_visualization,
    find_largest_obstacle,
    load_config,
    load_rectification_maps,
    rectify_pair,
)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--left-camera", type=int, default=2)
    parser.add_argument("--right-camera", type=int, default=0)
    parser.add_argument("--maps", type=Path, default=DEFAULT_MAPS)
    parser.add_argument("--config", type=Path, default=DEFAULT_CONFIG)
    parser.add_argument("--minimum-depth", type=float, default=10.0)
    parser.add_argument("--safe-distance", type=float, default=100.0)
    parser.add_argument("--minimum-area-fraction", type=float, default=0.01)
    parser.add_argument("--max-frames", type=int, default=0)
    parser.add_argument("--no-display", action="store_true")
    parser.add_argument("--output", type=Path)
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if args.left_camera == args.right_camera:
        raise ValueError("Left and right camera indices must be different.")
    if args.max_frames < 0:
        raise ValueError("--max-frames must be non-negative.")
    if args.no_display and args.max_frames == 0:
        raise ValueError("--no-display requires a positive --max-frames.")

    config = load_config(args.config)
    maps = load_rectification_maps(args.maps)
    matcher = create_matcher(config)
    left_camera = cv2.VideoCapture(args.left_camera)
    right_camera = cv2.VideoCapture(args.right_camera)
    if not left_camera.isOpened():
        raise RuntimeError(f"Could not open left camera {args.left_camera}.")
    if not right_camera.isOpened():
        left_camera.release()
        raise RuntimeError(f"Could not open right camera {args.right_camera}.")

    final_canvas = None
    frame_count = 0
    try:
        while True:
            if not left_camera.grab() or not right_camera.grab():
                raise RuntimeError("Could not grab synchronized stereo frames.")
            left_ok, left_color = left_camera.retrieve()
            right_ok, right_color = right_camera.retrieve()
            if not left_ok or not right_ok:
                raise RuntimeError("Could not retrieve synchronized stereo frames.")

            left_gray = cv2.cvtColor(left_color, cv2.COLOR_BGR2GRAY)
            right_gray = cv2.cvtColor(right_color, cv2.COLOR_BGR2GRAY)
            left_rectified, right_rectified = rectify_pair(
                left_gray, right_gray, maps
            )
            disparity = compute_disparity(
                left_rectified, right_rectified, config, matcher=matcher
            )
            depth = disparity_to_depth(disparity, config)
            _, obstacle = find_largest_obstacle(
                depth,
                min_depth=args.minimum_depth,
                max_depth=args.safe_distance,
                minimum_area_fraction=args.minimum_area_fraction,
            )

            final_canvas = cv2.remap(
                left_color,
                maps.left_x,
                maps.left_y,
                cv2.INTER_LANCZOS4,
                cv2.BORDER_CONSTANT,
            )
            if obstacle is None:
                cv2.putText(
                    final_canvas,
                    "SAFE",
                    (30, 60),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    1.4,
                    (0, 255, 0),
                    3,
                    cv2.LINE_AA,
                )
            else:
                x, y, width, height = obstacle.bounding_box
                cv2.rectangle(
                    final_canvas,
                    (x, y),
                    (x + width, y + height),
                    (0, 0, 255),
                    3,
                )
                cv2.putText(
                    final_canvas,
                    f"WARNING: {obstacle.mean_depth:.1f} cm",
                    (max(0, x), max(30, y - 10)),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.8,
                    (0, 0, 255),
                    2,
                    cv2.LINE_AA,
                )

            frame_count += 1
            if not args.no_display:
                cv2.imshow("Obstacle avoidance", final_canvas)
                cv2.imshow("Disparity", disparity_visualization(disparity))
                if cv2.waitKey(1) == 27:
                    break
            if args.max_frames and frame_count >= args.max_frames:
                break
    finally:
        left_camera.release()
        right_camera.release()
        if not args.no_display:
            cv2.destroyAllWindows()

    if final_canvas is None:
        raise RuntimeError("No stereo frames were processed.")
    if args.output:
        destination = args.output.expanduser().resolve()
        destination.parent.mkdir(parents=True, exist_ok=True)
        if not cv2.imwrite(str(destination), final_canvas):
            raise OSError(f"Could not write output image: {destination}")
        print(f"Saved final obstacle view: {destination}")
    print(f"Processed stereo frames: {frame_count}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Depth-Perception-Using-Stereo-Camera/cpp/obstacle_avoidance.cpp
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

#include "stereo_depth.hpp"

namespace {

struct Options {
    int leftCamera = 2;
    int rightCamera = 0;
    int maxFrames = 0;
    bool display = true;
    double minDepth = 10.0;
    double safeDistance = 100.0;
    double minimumAreaFraction = 0.01;
    std::filesystem::path maps =
        std::filesystem::path(STEREO_SOURCE_DIR) / "data" /
        "stereo_rectify_maps.xml";
    std::filesystem::path config =
        std::filesystem::path(STEREO_SOURCE_DIR) / "data" /
        "depth_estimation_params_cpp.xml";
    std::filesystem::path output;
};

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        auto requireValue = [&](const char* option) -> std::string {
            if (index + 1 >= argc) {
                throw std::invalid_argument(
                    std::string("Missing value for ") + option);
            }
            return argv[++index];
        };
        if (argument == "--left-camera") {
            options.leftCamera =
                std::stoi(requireValue("--left-camera"));
        } else if (argument == "--right-camera") {
            options.rightCamera =
                std::stoi(requireValue("--right-camera"));
        } else if (argument == "--maps") {
            options.maps = requireValue("--maps");
        } else if (argument == "--config") {
            options.config = requireValue("--config");
        } else if (argument == "--minimum-depth") {
            options.minDepth =
                std::stod(requireValue("--minimum-depth"));
        } else if (argument == "--safe-distance") {
            options.safeDistance =
                std::stod(requireValue("--safe-distance"));
        } else if (argument == "--minimum-area-fraction") {
            options.minimumAreaFraction =
                std::stod(requireValue("--minimum-area-fraction"));
        } else if (argument == "--max-frames") {
            options.maxFrames =
                std::stoi(requireValue("--max-frames"));
        } else if (argument == "--output") {
            options.output = requireValue("--output");
        } else if (argument == "--no-display") {
            options.display = false;
        } else {
            throw std::invalid_argument("Unknown option: " + argument);
        }
    }
    if (options.leftCamera == options.rightCamera) {
        throw std::invalid_argument(
            "Left and right camera indices must be different.");
    }
    if (options.maxFrames < 0) {
        throw std::invalid_argument(
            "--max-frames must be non-negative.");
    }
    if (!options.display && options.maxFrames == 0) {
        throw std::invalid_argument(
            "--no-display requires a positive --max-frames.");
    }
    return options;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const Options options = parseOptions(argc, argv);
        const auto config =
            learnopencv::stereo::loadConfig(options.config);
        const auto maps =
            learnopencv::stereo::loadRectificationMaps(options.maps);
        const auto matcher =
            learnopencv::stereo::createMatcher(config);

        cv::VideoCapture leftCamera(options.leftCamera);
        cv::VideoCapture rightCamera(options.rightCamera);
        if (!leftCamera.isOpened()) {
            throw std::runtime_error(
                "Could not open left camera " +
                std::to_string(options.leftCamera));
        }
        if (!rightCamera.isOpened()) {
            throw std::runtime_error(
                "Could not open right camera " +
                std::to_string(options.rightCamera));
        }

        cv::Mat finalCanvas;
        int frameCount = 0;
        while (true) {
            if (!leftCamera.grab() || !rightCamera.grab()) {
                throw std::runtime_error(
                    "Could not grab synchronized stereo frames.");
            }
            cv::Mat leftColor;
            cv::Mat rightColor;
            if (!leftCamera.retrieve(leftColor) ||
                !rightCamera.retrieve(rightColor) ||
                leftColor.empty() || rightColor.empty()) {
                throw std::runtime_error(
                    "Could not retrieve synchronized stereo frames.");
            }

            cv::Mat leftGray;
            cv::Mat rightGray;
            cv::cvtColor(
                leftColor, leftGray, cv::COLOR_BGR2GRAY);
            cv::cvtColor(
                rightColor, rightGray, cv::COLOR_BGR2GRAY);
            const auto [leftRectified, rightRectified] =
                learnopencv::stereo::rectifyPair(
                    leftGray, rightGray, maps);
            const cv::Mat disparity =
                learnopencv::stereo::computeDisparity(
                    leftRectified, rightRectified, config, matcher);
            const cv::Mat depth =
                learnopencv::stereo::disparityToDepth(
                    disparity, config);
            const auto [mask, obstacle] =
                learnopencv::stereo::findLargestObstacle(
                    depth, options.minDepth, options.safeDistance,
                    options.minimumAreaFraction);

            cv::remap(
                leftColor, finalCanvas, maps.leftX, maps.leftY,
                cv::INTER_LANCZOS4, cv::BORDER_CONSTANT);
            if (obstacle.has_value()) {
                cv::rectangle(
                    finalCanvas, obstacle->boundingBox,
                    cv::Scalar(0, 0, 255), 3);
                std::ostringstream label;
                label << "WARNING: " << std::fixed
                      << std::setprecision(1)
                      << obstacle->meanDepth << " cm";
                cv::putText(
                    finalCanvas, label.str(),
                    cv::Point(
                        std::max(0, obstacle->boundingBox.x),
                        std::max(30, obstacle->boundingBox.y - 10)),
                    cv::FONT_HERSHEY_SIMPLEX, 0.8,
                    cv::Scalar(0, 0, 255), 2, cv::LINE_AA);
            } else {
                cv::putText(
                    finalCanvas, "SAFE", cv::Point(30, 60),
                    cv::FONT_HERSHEY_SIMPLEX, 1.4,
                    cv::Scalar(0, 255, 0), 3, cv::LINE_AA);
            }

            ++frameCount;
            if (options.display) {
                cv::imshow("Obstacle avoidance", finalCanvas);
                cv::imshow(
                    "Disparity",
                    learnopencv::stereo::disparityVisualization(
                        disparity));
                if (cv::waitKey(1) == 27) {
                    break;
                }
            }
            if (
                options.maxFrames > 0 &&
                frameCount >= options.maxFrames) {
                break;
            }
        }
        if (options.display) {
            cv::destroyAllWindows();
        }
        if (finalCanvas.empty()) {
            throw std::runtime_error("No stereo frames were processed.");
        }
        if (!options.output.empty()) {
            if (!options.output.parent_path().empty()) {
                std::filesystem::create_directories(
                    options.output.parent_path());
            }
            if (!cv::imwrite(
                    options.output.string(), finalCanvas)) {
                throw std::runtime_error(
                    "Could not write obstacle output.");
            }
            std::cout << "Saved final obstacle view: "
                      << options.output << '\n';
        }
        std::cout << "Processed stereo frames: " << frameCount << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
